archunit/src/main/java/com/tngtech/archunit/core/domain/JavaCodeUnit.java

```diff
@@ -46,10 +47,14 @@
  * in particular every place, where Java code with behavior, like calling other methods or accessing fields, can
  * be defined.
  */
-public abstract class JavaCodeUnit extends JavaMember implements HasParameterTypes, HasReturnType, HasThrowsClause<JavaCodeUnit> {
-    private final JavaClass returnType;
+public abstract class JavaCodeUnit
+        extends JavaMember
+        implements HasParameterTypes, HasReturnType, HasTypeParameters<JavaCodeUnit>, HasThrowsClause<JavaCodeUnit> {
```

Comment:
I thought about it in the past and back the decided against it, because the covariant override seems to provide almost the same convenience, but the type signature is easier to understand (a recursive `SELF` type also doesn't come for free, because everyone using `JavaCodeUnit` will need to comprehend what that `<SELF>` means and deal with possible raw type warnings).
But yes, the cost is a more complex method signature in the super types, like `List<? extends JavaTypeVariable<? extends JavaCodeUnit>>` instead of `List<JavaTypeVariable<SELF>>`, since otherwise the covariant override doesn't work :disappointed:
But back then for annotations I decided it's still better to have some method signatures more complicated than necessary and the type signature simple vs. the other way round. In the end already `JavaMember` could have gotten the `SELF` type for `HasAnnotations<SELF>`.
I basically just guessed at that time, that most of the times users would

* call `getAnnotations()` on a `JavaMethod` or `JavaConstructor` where the signature is nice
* call other methods than `getAnnotations()` on `JavaMember`

Thus outweighing the cases where it is a little ugly :man_shrugging: